File: 백준/Silver/2644. 촌수계산/촌수계산.py
import sys
input = sys.stdin.readline


N = int(input()) # 정점 수
A, B = map(int, input().split()) # 구해야 하는 촌수 사람
M = int(input()) # 간선 수

arr = [[] for _ in range(N+1)] # 리스트 만들기
visited = [0] * (N+1) # 방문 여부이자 순서 저장
answer = [0] * (N+1)

for _ in range(M):
    a, b = map(int, input().split())
    arr[a].append(b) # 양방향 설정
    arr[b].append(a)

def dfs(x):
    visited[x] = 1 # 방문 체크
    for i in arr[x]: # 이어진 간선들 중
        if not visited[i]: # 방문하지 않았으면 탐색
            answer[i] = answer[x] + 1 # 촌수 +1
            dfs(i)

dfs(A)

if answer[B] == 0:
    print(-1)
else:
    print(answer[B])

File: 백준/Silver/2644. 촌수계산/촌수계산.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static int n;
    static boolean[][] family;
    static boolean[] visited;
    static int answer = Integer.MAX_VALUE;
    
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] target = new int[2];
        target[0] = Integer.parseInt(st.nextToken());
        target[1] = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(br.readLine());

        family = new boolean[n+1][n+1];
        visited = new boolean[n+1];
        for(int i=0; i<m; i++) {
            st = new StringTokenizer(br.readLine());
            int parent = Integer.parseInt(st.nextToken());
            int child = Integer.parseInt(st.nextToken());
            family[parent][child] = true;
            family[child][parent] = true;
        }

        visited[target[0]] = true;
        dfs(target[0], target[1], 0);

        if(answer == Integer.MAX_VALUE) System.out.println(-1);
        else System.out.println(answer);
    }

    private static void dfs(int now, int target, int cnt) {

        if(now == target) {
            answer = Math.min(cnt, answer);
            return;
        }
        
        for(int i=1; i<=n; i++) {
            if(family[now][i] && !visited[i]) {
                visited[i] = true;
                dfs(i, target, cnt+1);
                visited[i] = false;
            }
        }
    }
}